2021/BondielliCamarda/cusum.c

```diff
@@ -0,0 +1,213 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
+
+//structure that contains parameters for the algorithm
+struct ndpi_cusum_struct {
+    
+    //fixed parameters
+    struct {
+        double alpha, beta, sigma;
+        //maximum number of elements inside window
+        int window_size;
+    } params;
+
+    //array for read values
+    double *values_window;
+
+    //number of read values by the algorithm
+    u_int32_t num_values;
+    
+    //values calculated for the next iterations
+    double last_forecast, last_mu, square_sigma;
+
+    //indicates if the algorithm has started
+    bool start_cusum;
+};
+
+int ndpi_cusum_init(struct ndpi_cusum_struct *cusum, double alpha, double beta, double sigma, int window_size);
+
+int ndpi_cusum_add_value(struct ndpi_cusum_struct *cusum, const u_int64_t _value, double *forecast);
+
+double max(double a, double b);
+
+double mean(double *values, int size);
+
+int main() {
+    struct ndpi_cusum_struct cusum;
+    double v[] = {
+        31.908466339111,
+        87.339714050293,
+        173.47660827637,
+        213.92568969727,
+        223.32124328613,
+        230.60134887695,
+        238.09457397461,
+        245.8137512207,
+        251.09228515625,
+        251.09228515625,
+        259.21997070312,
+        261.98754882812,
+        264.78540039062,
+        264.78540039062,
+        270.47451782227,
+        173.3671875,
+        288.34222412109,
+        288.34222412109,
+        304.24795532227,
+        304.24795532227,
+        350.92227172852,
+        384.54431152344,
+        423.25942993164,
+        439.43322753906,
+        445.05981445312,
+        445.05981445312,
+        445.05981445312,
+        445.05981445312
+    };
+
+    int i, num = sizeof(v) / sizeof(double);
+    double alpha = 0.5, sigma = 10, beta = 0.95;
+
+    int ret = ndpi_cusum_init(&cusum, alpha, beta, sigma, 3);
+
+    printf("Start Alghoritm\n");
+
+    for (int i = 0; i < num; ++i) {
+        double prediction, threshold = 1500;
```

Comment:
Perche' la soglia e' statica e non dinamica come nell'exponential smoothing?

Reply:
Abbiamo preso spunto dal file dei colleghi ErminiLuziPinta e abbiamo fatto la stessa cosa.
Il nostro obiettivo era solo di dare un esempio d'utilizzo del calcolo di Cusum (infatti anche i valori letti dall'array sono casuali e presi da un suo esempio su ndpiReader.c).
Se vuole la possiamo fare dinamica (nel caso va bene calcolarla come nel Double Exponential Smoothing?).

Reply:
Io vorrei che facciate un passo in avanti rispetto a ErminiLuziPinta in quanto il loro codice era assai migliorabile. Fare la stessa cosa non aiuta a fare un passo in avanti.

Reply:
Abbiamo provato ad aggiungere la gestione della soglia dinamica (calcolandola come nelle slide).